### DISCUSSION TITLE
[3주차]_5장_안정 해시 설계

### DISCUSSION THREAD

**Discussion by minju412** [11th-SSAFY-19/large-scale-system-design#15] [General]
.

**Comment by chwangmin** [comment]
안정 해시 알고리즘에서 2^160의 크기를 가진 해시링이 존재한다고 했을 때, 해당 키에 대한 server를 찾는다면, 해당 키의 위치로부터 시계 방향으로 링을 탐색해 나가면서 만나는 첫 번째 서버다. 라고 하는데, 그러면 최악의 경우에서는 2^160 - 1번 확인할 수 있는 가능성이 있는 건가요?

**Reply by Hyunuk17** [reply]
가장 가까운 서버를 찾는 방법?
- 한칸 한칸 탐색인가?

- A: 트리 맵을 통해 서버의 위치를 탐색합니다

**Comment by minju412** [comment]
79페이지에서 "1번 서버가 죽으면 대부분 캐시 클라이언트가 데이터가 없는 엉뚱한 서버에 접속하게 된다는 뜻이다." 라는 문장이 있는데, `캐시 클라이언트`라는게 무엇인가요..? 검색해도 잘 나오지 않아 여쭤봅니다,,

**Reply by Hyunuk17** [reply]
A : 이번 주제가 N개의 캐시 서버를 분산하는 문제이기 때문에, 캐시에 접근하는 다양한 것들을 통칭하여 캐시 클라이언트라고 하고 있는 거 같습니다.

**Comment by minju412** [comment]
`안정 해시 알고리즘의 기본 절차` 중 "서버와 키를 균등 분포 해시 함수를 사용해 해시 링에 배치한다." 라는 내용이 있는데요,
그 아래 `기본 구현법의 두 가지 문제`에서 "키의 균등 분포를 달성하기가 어렵다."라는 내용이 있습니다.

이 둘이 상충되는 내용이라고 생각하는데 어떻게 생각하시나요?

> 84페이지

**Reply by Hyunuk17** [reply]
추가되거나 삭제되는 상황을 가정하면 입니다..

**Comment by minju412** [comment]
SHA-1 해시 함수가 지금은 잘 사용되지 않는 이유가 무엇인가요? -혁진-

**Reply by Hyunuk17** [reply]
A: 
- 컴퓨터의 연산 능력 발전에 따라 비교적 취약한 알고리즘
- SHA-1 : 160bit
- SHA-256 : 256bit

**Reply by chwangmin** [reply]
논외) https://m.blog.naver.com/iank2016/222380554499
2008년도 SHA-1번 충돌이 실제로 발생했다.

**Comment by minju412** [comment]
핫스팟 키 문제에서 안정해시를 통해 데이터를 더 균등하게 분배한다고 했는데, 어떤 방법으로 그럴 수 있나요?

**Reply by Hyunuk17** [reply]
A: 가상 노드(Virtual Node)를 활용하여 여러 파티션으로 분할함으로서 특정 파티션에 부하가 집중되지 않고 키가 더 균등하게 분배될 수 있습니다

**Comment by minju412** [comment]
핫스팟 키 문제가 발생하면, 트래픽이 몰리는 값을 dba가 수동으로 바꿔줄 수 있나요? -광민-

**Reply by Hyunuk17** [reply]
해시 함수 f에 의해 할당된 값을 가장 가까운 서버에 할당하는 방법으로 진행되기 때문에, 특정 키만 연결된 서버를 바꾸려면 그 키만 별도의 해시 함수를 적용해야 할 텐데 말이 안될 것 같습니다.

대신 트래픽이 몰리는 파티션에 서버를 추가하여 파티션 자체를 분할하여 부하를 줄이는 방법이 가능할 것 같습니다.

**Comment by minju412** [comment]
마이클잭슨에 몰리는 트래픽을 반반 나눠서 두 서버로 나누는 것인지 vs
마이클잭슨과 저스틴비버처럼 트래픽이 많은 값을 떼어놓자는 것인가요? -혁진-

**Reply by minju412** [reply]
요청을 분산하는게 아니라 데이터를 분산하는 것 같다. -현재-

**Reply by Hyunuk17** [reply]
가상 노드는 파티션을 여러 개로 분할하여 데이터가 하나의 물리적 노드에 집중되는 것을 막기 위해 도입한 것으로, 하나의 키 값에 대한 처리보다는 여러 키의 균등한 분배를 목적으로 하고 있습니다.
